### PR TITLE
Integrate palette inferrer

### PR DESCRIPTION
**Pull request by jamieparkinson**
Does what it says. Terraform PR to come separately because there is already quite a lot of this.

### CODE REVIEW

**Review by harrisonpim — APPROVED** [review]
hoping that this can be quickly followed up with #811 after a few small tweaks

File: common/elasticsearch/src/test/scala/uk/ac/wellcome/elasticsearch/ImagesIndexConfigTest.scala [hunk 1]
```diff
@@ -39,24 +39,33 @@ class ImagesIndexConfigTest
     }
   }
 
-  it("cannot ingest an image with image vectors that are shorter than 2048") {
+  it("cannot ingest an image with image vectors that are longer than 2048") {
```
Comment (kenoir) [inline common/elasticsearch/src/test/scala/uk/ac/wellcome/elasticsearch/ImagesIndexConfigTest.scala:42]: Assume this is fixing a typo.

File: pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/adapters/InferrerAdapter.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport._
+import grizzled.slf4j.Logging
+import io.circe.Decoder
+import uk.ac.wellcome.models.work.internal.InferredData
+import uk.ac.wellcome.platform.inference_manager.models.DownloadedImage
+
+import scala.concurrent.Future
+
+/*
+ * An InferrerAdapter is specific to the inferrer and the data that is being augmented.
```
Comment (kenoir) [inline pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/adapters/InferrerAdapter.scala:15]: Useful comment.

File: pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala [hunk 3]
```diff
@@ -38,48 +50,94 @@ class InferenceManagerWorkerService[Destination](
...
         .via(sendAugmented)
         .asSource
         .map { case (_, msg) => msg }
     )
 
-  private def createRequest[Ctx] =
-    FlowWithContext[DownloadedImage, Ctx].map { image =>
-      (inferrerAdapter.createRequest(image), image)
+  private def createRequests[Ctx] =
+    FlowWithContext[DownloadedImage, Ctx].mapConcat { image =>
```
Comment (kenoir) [inline pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala:63]: `FlowWithContext` ... new to me, very handy.

**Review by kenoir — APPROVED** [review]

File: pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/adapters/InferrerAdapter.scala [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ * Implementors must provide:
+ * - The type of the inferrer response
+ * - A Decoder for that response
+ * - A function to create an HTTP request from the input data type
+ * - A function to augment input data with a response, returning the output data type
+ *
+ * Additionally, the trait provides the logic for handling different HTTP response statuses
+ */
+trait InferrerResponse
```
Comment (warrd) [inline pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/adapters/InferrerAdapter.scala:24]: I don't see the need of having this trait personally, as it's not a sealed trait allowing matching or an interface which subtypes have to fulfill. Would be just as clear with `type Response` below imo

File: pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala [hunk 5]
```diff
@@ -3,31 +3,43 @@ package uk.ac.wellcome.platform.inference_manager.services
...
 import uk.ac.wellcome.typesafe.Runnable
 
 import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
 import scala.util.{Failure, Success, Try}
 
+case class AdapterResponseBundle[ImageType](
+  image: ImageType,
+  adapter: InferrerAdapter,
+  response: Try[InferrerResponse]
```
Comment (warrd) [inline pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala:27]: Related to my comment above, there is nothing you can do with the success type of a `Try[InferrerResponse]` as the trait does provide any interface. Are you planning to add some required functionality to the trait so the response can be used? If not I see 2 options:

* Making this an error instead (e.g. `error: Option[Exception]` or something)
* Adding another type parameter to the case class so we can use the concrete type: `AdapterResponseBundle[ImageType, Response]`

File: pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala [hunk 6]
```diff
@@ -38,48 +50,94 @@ class InferenceManagerWorkerService[Destination](
...
+    FlowWithContext[AdapterResponseBundle[DownloadedImage], Ctx]
+      .via {
+        Flow[(AdapterResponseBundle[DownloadedImage], Ctx)]
+          .groupBy(parallelism, _ match {
+            case (_, msg) => msg.messageId()
+          }, allowClosedSubstreamRecreation = true)
+          .groupedWithin(inferrerAdapters.size, maxInferrerWait)
+          .map { elements =>
+            elements.filter {
+              case (AdapterResponseBundle(_, _, Failure(_)), _) => false
```
Comment (warrd) [inline pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala:112]: Don't understand all the requirements, but do we need to retry failures or anything, or is it fine just ignoring them like this?

File: pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala (hunk 3 above)
Comment (jamieparkinson) [reply pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala:63]: As with lots of things in akka-streams it has some non-obvious restrictions in order to keep it theoretically pure, but on the whole it definitely is handy

File: pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala [hunk 7]
```diff
@@ -38,48 +50,94 @@ class InferenceManagerWorkerService[Destination](
...
+                s"Did not receive responses from $failedInferrers within $maxInferrerWait")
+            }
+            elements
+          }
+          .map { elements =>
+            val inferredData = elements.foldLeft(InferredData.empty) {
+              case (
+                  data,
+                  (AdapterResponseBundle(_, adapter, Success(response)), _)) =>
+                adapter.augment(data, response.asInstanceOf[adapter.Response])
```
Comment (warrd) [inline pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala:131]: Ah in light of my comment above a parameterised type would be best for the `Try` argument so we can avoid a type cast here.

**Review by warrd — COMMENTED** [review]
Just saw you merged this, have some comments you may / may not want to address in a subsequent PR.

**Pull request merged**